libraries/node-core-library/src/Executable.ts

```diff
@@ -0,0 +1,419 @@
+  | NodeJS.WritableStream | NodeJS.ReadableStream
+  | number | undefined;
+
+/**
+ * Typings for IExecutableSpawnSyncOptions.stdio.
+ * @beta
+ */
+export type ExecutableStdioMapping = 'pipe' | 'ignore' | 'inherit' | ExecutableStdioStreamMapping[];
+
+/**
+ * Options for Executable.tryResolve().
+ * @beta
+ */
+export interface IExecutableResolveOptions {
+  /**
+   * The current working directory.  If omitted, process.cwd() will be used.
+   */
+  currentWorkingDirectory?: string;
+
+  /**
+   * The environment variables for the child process.  If omitted, process.env will be used.
+   */
+  environment?: NodeJS.ProcessEnv;
+}
+
+/**
+ * Options for Executable.execute().
+ * @beta
+ */
+export interface IExecutableSpawnSyncOptions extends IExecutableResolveOptions {
+  /**
+   * The content to be passed to the child process's stdin.
+   */
+  input?: string;
+
+  /**
+   * The stdio mappings for the child process.
+   */
+  stdio?: ExecutableStdioMapping;
+
+  /**
+   * The maximum time the process is allowed to run before it will be terminated.
+   */
+  timeoutMs?: number;
+
+  /**
+   * The largest amount of bytes allowed on stdout or stderr for this synchonous operation.
+   * If exceeded, the child process will be terminated.  The default is 200 * 1024.
+   */
+  maxBuffer?: number;
+
+  /**
+   * The encoding used for all stdio inputs/outputs.  The default is 'buffer'.
+   */
+  encoding?: string | BufferEncoding;
+}
+
+// Common environmental state used by Executable members
+interface IExecutableContext {
+  currentWorkingDirectory: string;
+  environment: NodeJS.ProcessEnv;
+  // For Windows, the parsed PATHEXT environment variable
+  windowsExecutableExtensions: string[];
+}
+
+/**
+ * The Executable class provides a safe, portable, recommended solution for tools that need
+ * to launch child processes.
+ *
+ * @remarks
+ * The child_process API lacks certain higher-level features, which encourages reliance on
+ * the operating system shell.  Invoking the OS shell is not safe, not portable, and
+ * generally not recommended:
+ *
+ * - Different shells have different behavior and command-line syntax, and which shell you
+ *   will get is unpredictable.  There is no universal shell guaranteed to be available on
+ *   all platforms.
+ *
+ * - If a command parameter contains symbol characters, a shell may process them, which
+ *   can introduce a security vulnerability
+ *
+ * - Each shell has different rules for escaping these symbols.  On Windows, the default
+ *   shell is incapable of escaping certain character sequences.
+ *
+ * The Executable class provides a pure JavaScript implementation of the primitive functionality
+ * for searching the PATH, correctly passing command-line arguments (which may contain
+ * symbols or spaces), and finding the registered launcher based on the file extension
+ * or POSIX shebang.  This primitive functionality is sufficient (and recommended) for
```

Comment:
Do we want to include this here, or just in Rushell?

Reply:
Perhaps this should say "and finding the registered launcher based on the file extension on Windows,  or POSIX shebang on POSIX"

---
In reply to: [208418289](https://github.com/Microsoft/web-build-tools/pull/764#discussion_r208418289) [](ancestors = 208418289)